xrootd/client.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package xrootd // import "go-hep.org/x/hep/xrootd"
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+
+	"go-hep.org/x/hep/xrootd/encoder"
+	"go-hep.org/x/hep/xrootd/streammanager"
+)
+
+var logger = log.New(os.Stderr, "xrootd: ", log.LstdFlags)
+
+// A Client to xrootd server
+type Client struct {
+	connection      *net.TCPConn
+	sm              *streammanager.StreamManager
+	protocolVersion int32
+}
+
+type serverError struct {
+	Code    int32
+	Message string
+}
+
+func (err serverError) Error() string {
+	return fmt.Sprintf("Server error %d: %s", err.Code, err.Message)
+}
+
+const responseHeaderSize = 2 + 2 + 4
+
+type responseHeader struct {
+	StreamID   streammanager.StreamID
+	Status     uint16
+	DataLength int32
+}
+
+// New creates a client to xrootd server at address
+func New(ctx context.Context, address string) (*Client, error) {
+	conn, err := createTCPConnection(address)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{conn, streammanager.New(), 0}
+
+	go client.consume()
```

Comment:
this goroutine is leaked: there is no way one can properly close it nor reclaim its resources.
perhaps it should be passed a context value.

naively, when I saw the context value being passed to the `New` function, I had expected it would be bound to the lifetime of the client.